### PR TITLE
feat(connect): add parquet support

### PR DESCRIPTION
**Pull request by andrewgazelka**

### CODE REVIEW

**Comment by andrewgazelka** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3360?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3360** <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3360?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3360?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#3355** <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3355?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 3 other dependent PRs ([#3350](https://github.com/Eventual-Inc/Daft/pull/3350) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3350?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3354](https://github.com/Eventual-Inc/Daft/pull/3354) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3354?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3357](https://github.com/Eventual-Inc/Daft/pull/3357) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3357?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3342** <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3342?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 6 other dependent PRs ([#3344](https://github.com/Eventual-Inc/Daft/pull/3344) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3344?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3345](https://github.com/Eventual-Inc/Daft/pull/3345) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3345?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3348](https://github.com/Eventual-Inc/Daft/pull/3348) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3348?utm_source=stack-comment-icon" target="_blank">Graphite</a> and 3 others)
* **#3343** <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3343?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#3309](https://github.com/Eventual-Inc/Daft/pull/3309) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3309?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3326](https://github.com/Eventual-Inc/Daft/pull/3326) <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3326?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3341** <a href="https://app.graphite.dev/github/pr/Eventual-Inc/Daft/3341?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/daft-connect/src/op/execute/write.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            let mut result_stream = native_executor
+                .run(plan.psets, cfg.into(), None)?
+                .into_stream();
+
+            // this is so we make sure the operation is actually done
+            // before we return
+            //
+            // an example where this is important is if we write to a parquet file
+            // and then read immediately after, we need to wait for the write to finish
+            while let Some(_result) = result_stream.next().await {}
```
Comment (colin-ho) [inline src/daft-connect/src/op/execute/write.rs:131]: Why not return the results of the write?

File: tests/connect/test_parquet.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        # Create DataFrame from range(10)
+        df = spark_session.range(10)
+
+        # Write DataFrame to parquet directory
+        parquet_dir = os.path.join(temp_dir, "test.parquet")
+        df.write.parquet(parquet_dir)
+
+        # List all files in the parquet directory
+        parquet_files = [f for f in os.listdir(parquet_dir) if f.endswith(".parquet")]
+        print(f"Parquet files in directory: {parquet_files}")
```
Comment (colin-ho) [inline tests/connect/test_parquet.py:21]: Print statement

File: tests/connect/test_parquet.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+from __future__ import annotations
+
+import os
+import shutil
+import tempfile
+
+
+def test_write_parquet(spark_session):
+    # Create a temporary directory
+    temp_dir = tempfile.mkdtemp()
```
Comment (colin-ho) [inline tests/connect/test_parquet.py:10]: You can just use `with tempfile.TemporaryDirectory() as dirname:` then you don't need to implement cleanup yourself.

File: src/daft-connect/src/op/execute/write.rs (hunk 1 above)
Comment (andrewgazelka) [reply src/daft-connect/src/op/execute/write.rs:131]: because a write is not a read?

why would we want to do this?

Comment (colin-ho) [reply src/daft-connect/src/op/execute/write.rs:131]: It returns the file names of the written files.

Comment (andrewgazelka) [reply src/daft-connect/src/op/execute/write.rs:131]: https://github.com/Eventual-Inc/Daft/issues/3550

I want to unify these then. made a separate issue. can we resolve and merge? and then do that in another PR?

**Review by colin-ho — APPROVED** [review]

File: src/daft-connect/src/op/execute/write.rs (hunk 1 above)
Comment (colin-ho) [reply src/daft-connect/src/op/execute/write.rs:131]: Sounds good

**Pull request merged**